---
title: "01_process_idats"
author: "Puvvula"
date: "2023-06-30"
output: pdf_document
---

```{r}
library(pacman)
pacman::p_load(tidyverse, janitor, stringi, UpSetR)
pheno <- "~/Documents/methylation_pilot/"
idats<- "~/Documents/methylation_pilot/idat_files/"
processed<- "~/Documents/methylation_pilot/processed/"
```

#read chip and de-identified subject info
```{r}
pheno <- read_csv(paste0(pheno, "meth_pheno.csv")) |>
  clean_names() |>
  mutate(gsm_id = paste(sentrix_id, sentrix_position, sep = "_")) 

unique_counts <- pheno |>
  group_by(sample_type) |>
  summarise(unique_participants = n_distinct(participant_id),
            n_chips = n_distinct(sentrix_id))
```

#visualize chip info
```{r}
# Generate the heatmap using ggplot2
cb_palette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7",
                   "#999999", "#A69F00", "#6AB4E9", "#109E73", "#E0E442", "#4072B2", "#B55E00", "#EC79A7")

ggplot(pheno, aes(x = sample_type, y = participant_id, fill = factor(sentrix_id))) +
  geom_tile(width = 0.5, height = 0.8,) +
  scale_x_discrete(expand = c(0, 0))+
  scale_fill_manual(values = cb_palette)+
  labs(x = "Sample Type", y = "Subject ID", fill = "Sentry ID") +
  theme_minimal()+
  guides(fill = "none")
```

#================= Mehtylation data ======================#
```{r}
pacman::p_load(minfi, ENmix, IlluminaHumanMethylation450kmanifest, 
               IlluminaHumanMethylation450kanno.ilmn12.hg19, FlowSorted.Blood.450k, 
               sva, wateRmelon, RColorBrewer)
```

#read idats
```{r}
#select one type of sample to process
pheno_by_tiss <- pheno |> filter(sample_type == "CBMC")

#idat file names
idat_files<-  paste0(idats, pheno_by_tiss$gsm_id)

# get the 450k annotation data
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)

#read raw idats
rgSet = read.metharray(basenames=idat_files)
```

#quality check
```{r}
# calculate the detection p-values
detP <- detectionP(rgSet)

# examine mean detection p-values across all samples to identify any failed samples
pal <- brewer.pal(8,"Dark2")
barplot(colMeans(detP), col=pal[factor(pheno_by_tiss$sentrix_id)], las=2, 
        cex.names=0.8, ylab="Mean detection p-values")

```

#minfi qcReport
```{r}
#get minfi qc report
qcReport(rgSet, sampNames=pheno_by_tiss$sentrix_id, sampGroups=pheno_by_tiss$sentrix_id, 
         pdf="qcReport.pdf")
```

# remove poor quality samples
```{r}
# Drop samples if 5% of probes are above the p-value threshold
samples_to_remove <- colSums(detP > 1e-7) > (0.05 * nrow(rgSet))
table(samples_to_remove)

# Remove samples exceeding the threshold from the methylation data
rgset_f2 <- rgSet[!samples_to_remove, ]
```

#generate methyl and genomic ratio set
```{r}
# normalize the data; this results in a GenomicRatioSet object
mSetSq_noob <- preprocessNoob(rgset_f2) 
mSetSq_bmiq <- BMIQ(mSetSq_noob)

#convert preprocessed +BMIQ object to a genomic ratio set
mSetSq_v1<- makeGenomicRatioSetFromMatrix(mSetSq_bmiq)

# create a MethylSet object from the raw data for plotting
mSetRaw <- preprocessRaw(rgset_f2)
```

#multi dimensional scaling plots
#principle components analysis and are an unsupervised method for looking at the similarities and differences between the various samples. Samples that are more similar to each other should cluster together, and samples that are very different should be further apart on the plot. Dimension one (or principle component one) captures the greatest source of variation in the data, dimension two captures the second greatest source of variation in the data and so on.  Colouring the data points or labels by known factors of interest can often highlight exactly what the greatest sources of variation are in the data. It is also possible to use MDS plots to decipher sample mix-ups.
```{r}
#change dim parameter to select different PCs from PCA
par(mfrow=c(1,3))
plotMDS(getM(mSetSq), top=1000, gene.selection="common", 
        col=pal[factor(pheno_by_tiss$sentrix_id)], dim=c(1,3))

plotMDS(getM(mSetSq), top=1000, gene.selection="common", 
        col=pal[factor(pheno_by_tiss$sentrix_id)], dim=c(2,3))

plotMDS(getM(mSetSq), top=1000, gene.selection="common", 
        col=pal[factor(pheno_by_tiss$sentrix_id)], dim=c(3,4))
```

#removing probes with poor quality using a p-value threshold
```{r}
# ensure probes are in the same order in the mSetSq and detP objects
detP <- detP[match(featureNames(mSetSq_v1),rownames(detP)),] 

# remove any probes that have failed in one or more samples
keep <- rowSums(detP < 1e-7) == ncol(mSetSq_v1) 
table(keep)

mSetSqFlt <- mSetSq_v1[keep,]
dim(mSetSqFlt)
```

#remove probes on sex chromosome
```{r}
# if your data includes males and females, remove probes on the sex chromosomes
keep <- !(featureNames(mSetSqFlt) %in% ann450k$Name[ann450k$chr %in% 
                                                        c("chrX","chrY")])
table(keep)
mSetSqFlt <- mSetSqFlt[keep,]
dim(mSetSqFlt)
```

# exclude cross reactive probes
# cross-hybrid probe list from: https://www.tandfonline.com/doi/full/10.4161/epi.23470
```{r}
xReactiveProbes <- read_csv(url("https://epigen.ccm.sickkids.ca/sample-report/data/quality_control/cross_reactive_probes.csv"),
                             col_names = c("TargetID"))

keep <- !(rownames(mSetSqFlt) %in% xReactiveProbes$TargetID)
table(keep)

mSetSqFlt <- mSetSqFlt[keep,] 
dim(mSetSqFlt)
```

# remove probes with SNPs at CpG site
```{r}
mSetSqFlt <- dropLociWithSnps(mSetSqFlt)
dim(mSetSqFlt)
```

#checking the PCA plots again
```{r}
#change dim parameter to select different PCs from PCA
par(mfrow=c(1,3))
plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", 
        col=pal[factor(pheno_by_tiss$sentrix_id)], dim=c(1,3))

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", 
        col=pal[factor(pheno_by_tiss$sentrix_id)], dim=c(2,3))

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", 
        col=pal[factor(pheno_by_tiss$sentrix_id)], dim=c(3,4))
```

#beta density plots
```{r}
par(mfrow=c(1,3))
densityPlot(getBeta(mSetRaw), sampGroups = pheno$sentrix_id, legend = F, main="Raw methylset")
densityPlot(getBeta(mSetSq_noob), sampGroups = pheno$sentrix_id, legend = F, main="Noob preprocessed \n+BMIQ methylset")
densityPlot(getBeta(mSetSqFlt), sampGroups = pheno$sentrix_id, legend = F, main="Noob preprocessed + \nBMIQ +filtered methylset")
```

# calculate M-values and beta for statistical analysis
```{r}
mVals <- getM(mSetSqFlt)

bVals <- getBeta(mSetSqFlt)

```

#plot m and beta values
```{r}
par(mfrow=c(1,2))
densityPlot(bVals, sampGroups=pheno_by_tiss$sentrix_id, main="Beta values", 
            legend=FALSE, xlab="Beta values")
densityPlot(mVals, sampGroups=pheno_by_tiss$sentrix_id, main="M-values", 
            legend=FALSE, xlab="M values")
```


#reference based cell count estimation from rgset for cord blood
```{r}
minfi.LC<- estimateCellCounts(rgSet,compositeCellType = "CordBlood", 
                              meanplot=TRUE, cellTypes = c("CD8T","CD4T", "NK","Bcell","Mono","Gran","nRBC"),
                              returnAll = FALSE, meanPlot = TRUE, verbose = TRUE, processMethod="preprocessNoob") |>
  as.data.frame() |>
  rownames_to_column("gsm_id")

```

#export cell type counts
```{r}
cbmc_pheno <- right_join(minfi.LC, pheno_by_tiss, by = "gsm_id")
```

#export rda objects for downstream analysis
```{r}
save(bVals, mVals, mSetSqFlt, cbmc_pheno, file = paste0(processed, "cbmc_processed.rda"))
```


#=======================================



